linkerd fails to install when removing port 443 from ignoreOutboundPorts

jimil749 opened the pull request:
Fixes #5616  

Changed control-plane templates i.e controller, identity, destination, proxy-injector and sp-validator to add 443 to skip ports if not present. This prevents failing of linkerd installation when 443 is not present in ignoreOutboundPorts.

PS: I am a bit new to helm-templates so I'm not totally sure about this one but this does add "443" to the list of outbound skip ports in the control plane resource template
Signed-off-by: Jimil Desai <jimildesai42@gmail.com>

<!--  Thanks for sending a pull request!

If you already have a well-structured git commit message, chances are GitHub
set the title and description of this PR to the git commit message subject and
body, respectively. If so, you may delete these instructions and submit your PR.

If this is your first time, please read our contributor guide:
https://github.com/linkerd/linkerd2/blob/main/CONTRIBUTING.md

The title and description of your Pull Request should match the git commit
subject and body, respectively. Git commit messages are structured as follows:

```
Subject

Problem

Solution

Validation

Fixes #[GitHub issue ID]

DCO Sign off
```

Example git commit message:

```
Introduce Pull Request Template

GitHub's community guidelines recommend a pull request template, the repo was
lacking one.

Introduce a `PULL_REQUEST_TEMPLATE.md` file.

Once merged, the
[Community profile checklist](https://github.com/linkerd/linkerd2/community)
should indicate the repo now provides a pull request template.

Fixes #3321

Signed-off-by: Jane Smith <jane.smith@example.com>
```

Note the git commit message subject becomes the pull request title.

For more details around git commits, see the section on Committing in our
contributor guide:
https://github.com/linkerd/linkerd2/blob/main/CONTRIBUTING.md#committing
-->

kleimkuhler commented:
@jimil749 just a heads up it looks like some tests need to be fixed for this!

jimil749 commented:
@kleimkuhler, I noticed! I was having some difficulty in running tests locally, so I'll resolve that first, and fix the test ASAP. Is that fine? Also, do you think the approach here is fine? :sweat_smile:

kleimkuhler commented:
@jimil749 Yep this seems correct so far. As already mentioned in the issue, I agree this fits in well as part of the Helm templating only in control plane templates.

jimil749 commented:
Hi @kleimkuhler ! Sorry for the delay, was caught up in some other work. I've updated the helm template to remove the `global` field (as per the latest commit!) and also added `443` port in the need test cases. But somehow the tests still fail and I'm not sure why? Could you help me figure out the reason?

kleimkuhler commented:
@jimil749 No problem! The unit tests are failing because the golden files expect different output. With this change, the output of `install` is different, and therefore it no longer matches what is in the golden files.

To update the golden files automatically, you can run `go test ./... --update`. You will see it change a lot of the files and that will fix the unit tests.

jimil749 commented:
Thanks for the help, @kleimkuhler! :) 
Running `go test ./... --update` fixed the unit tests!

kleimkuhler reviewed (approved):
This looks good! I'd like @alpeb or @Pothulapati who are more familiar with the Helm templating to take a look as well.

alpeb commented on charts/linkerd2/templates/controller.yaml:
```diff
@@ -128,7 +128,10 @@ spec:
-      - {{- include "partials.proxy-init" . | indent 8 | trimPrefix (repeat 7 " ") }}
+      {{- if not (contains "443" ( .Values.proxyInit.ignoreOutboundPorts | toString ))}}
+      {{- $_ := set $tree.Values.proxyInit "ignoreOutboundPorts" (printf "%s,443" ( .Values.proxyInit.ignoreOutboundPorts | toString )) -}}
+      {{- end}}      
```
There are extra spaces at the end of this line that explain the diffs in the golden files. It seems the other templates except for identity.yaml do as well. Also, to avoid repetition, IMO it'd make sense to move this logic into `_proxy-init.tpl` itself.

alpeb replied:
Please disregard my comment about moving this into `_proxy-init.tpl`. I realize now we only want this for the control-plane deployments... So the only thing left is to remove those spaces :wink:

jimil749 replied:
Done, @alpeb!
